packages/babel-plugin-proposal-do-expressions/src/index.ts

```diff
@@ -10,19 +12,194 @@ export default declare(api => {
     visitor: {
       DoExpression: {
         exit(path) {
-          const { node } = path;
-          if (node.async) {
+          if (path.node.async) {
             // Async do expressions are not yet supported
             return;
           }
-          const body = node.body.body;
-          if (body.length) {
-            path.replaceExpressionWithStatements(body);
-          } else {
-            path.replaceWith(path.scope.buildUndefinedNode());
-          }
+          transformDoExpression(path);
         },
       },
     },
   };
+
+  function transformDoExpression(path: NodePath) {
+    do {
+      if (!path.parentPath) {
+        throw new Error("DoExpression must be in a statement");
+      }
+      if (Array.isArray(path.container) && path.isStatement()) {
+        // Flatten the closest parent statement
+        flattenStatement(path);
+        break;
+      }
+      if (path.isArrowFunctionExpression()) {
+        const body = path.get("body");
+        body.assertExpression();
```

Comment:
A do expression could be a parameter initializer or a computed key, 
```js
({ [do {'key'}]: x } = do { value }) => {}
```
in either case this assertion may fail.